core/chains/evm/logpoller/log_poller.go

```diff
@@ -305,8 +305,28 @@ func (lp *logPoller) Filter(from, to *big.Int, bh *common.Hash) ethereum.FilterQ
 // Replay signals that the poller should resume from a new block.
 // Blocks until the replay is complete.
 // Replay can be used to ensure that filter modification has been applied for all blocks from "fromBlock" up to latest.
-func (lp *logPoller) Replay(ctx context.Context, fromBlock int64) error {
-	latest, err := lp.ec.HeadByNumber(ctx, nil)
+func (lp *logPoller) Replay(ctx context.Context, fromBlock int64, async bool) error {
+	replayCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		// merge log poller context and parent context.  If either of them are cancelled, abort replay
+		// This is so that lp.run() can pass a single context to lp.PollAndSaveLogs().  Not used in this
+		// function since we want to return differentiate between the two, for returning proper error code
+		select {
+		case <-lp.ctx.Done():
+			cancel()
+		case <-ctx.Done():
+			if !async {
+				// In the async case, we don't want to cancel the replay after we've returned to the caller (unless the
+				// log poller is shutting down).  This is important for a web controller, since it will always
+				// cancel the context as soon as the request is complete, which will likely be before the replay is done.
```

Comment:
But context cancellation isn't prevented by avoiding this particular call, because the `replayCtx` is derived directly from the original `ctx` argument and will automatically be cancelled.

Reply:
Ah!  I was wondering why these tests weren't passing.  Kept tweaking the tests today thinking it was a problem with them, but you're right this is a bug... in fact, so is the `defer cancel()` line, for the same reason.  Thanks for looking at it, guess I need to refactor some things.